### PR TITLE
DM-7041: Update post-qa for validate_base-type JSON

### PR DESCRIPTION
**Pull request by jonathansick**
- Updates the JSON shims for work with validate_base as of [DM-7042](https://jira.lsstcorp.org/browse/DM-7042).
- Includes JSON Schema-based testing. See `postqa/schemas/squash.json` and the `postqa.schemas` module. Unit validation schemas for sub-documents are dynamically created from the 'master' schema.
- Improve test coverage to 92%, including mocking Git repos in `lsstsw`.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import os
+import io
+import json
+import pytest
+
+
+@pytest.fixture
+def lsstsw_dir():
+    """A minimal mock lsstsw/ installation is in /test_data/lsstsw."""
```
Comment (jmatt) [inline tests/conftest.py]: In the comment it should be "data" instead of "test_data".

File: tests/test_pkgdata.py [hunk 2]
```diff
@@ -6,36 +6,19 @@
...
-
 def test_manifest(manifest):
+    """Test Manifest's parsing against manually-extracted data."""
     m = Manifest(manifest)
 
     job_json = m.json
     assert 'packages' in job_json
     assert isinstance(job_json['packages'], list)
 
+    # Sample data extracted from test_data/lsstsw/build/manifest/.txt
```
Comment (jmatt) [inline tests/test_pkgdata.py]: In the comment it should be "tests" not "test_data" I believe.

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+"""Common pytest fixtures."""
+
+from __future__ import (division, absolute_import, print_function,
```
Comment (jmatt) [inline tests/conftest.py:3]: There are similar boilerplate includes in a few of the python modules. This isn't wrong but it an be misleading. For instance I was wondering where future print and division were being used in this module.

Do we, as a team, have an opinion on this?

File: tests/test_jsonshim.py [hunk 4]
```diff
@@ -6,56 +6,14 @@
...
-def test_shim_validate_drp(vdrp_cfht_output_r):
+def test_measurements_schema(vdrp_cfht_output_r):
+    """Validate the schema of `measurements` json sub-document."""
     job_json = shim_validate_drp(vdrp_cfht_output_r)
-    assert 'measurements' in job_json
-    assert len(job_json['measurements']) == 3
-    for measurement in job_json['measurements']:
-        assert 'metric' in measurement
-        assert 'value' in measurement
-
```
Comment (jmatt) [inline tests/test_jsonshim.py:48]: There may be value in testing failure conditions and any assumed edge cases too. The tests are relatively easy to create and it'll verify that it is in fact failing when we assume it should.

**Review by jmatt — APPROVED** [review]

**Comment by jmatt** [comment]
:shipit:

File: tests/conftest.py (hunk 3 above)
Comment (jonathansick) [reply tests/conftest.py:3]: Hmm. I dunno. I do this for all my Python 2.7+3 cross-compatible projects just so that I don't get surprises later.

File: tests/test_jsonshim.py (hunk 4 above)
Comment (jonathansick) [reply tests/test_jsonshim.py:48]: Good point.

**Comment by jonathansick** [comment]
Thanks @jmatt, I've tested more edge cases for validation failures. Should be good now.
